Fix file capabilites droping in Dockerfile

abdulrahimiliasu opened the pull request:
fixes #42655 

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
moved copyXattr function out of doCopyXattrs function, so that security capabilities are copied
**- How I did it**
doCopyXattrs() never reached due to copyXattrs boolean being false, as a result file capabilities not being copied.

**- How to verify it**

***Test Case***

```dockerfile
FROM registry1-docker-io.repo.lab.pl.alcatel-lucent.com/library/alpine:latest
RUN apk --no-cache add libcap && setcap cap_net_admin=eip /sbin/apk
RUN setcap -v cap_net_admin=eip /sbin/apk
```

***Test Result***
```
Sending build context to Docker daemon  2.048kB
Step 1/3 : FROM registry1-docker-io.repo.lab.pl.alcatel-lucent.com/library/alpine:latest
latest: Pulling from library/alpine
a0d0a0d46f8b: Pull complete
Digest: sha256:e1c082e3d3c45cccac829840a25941e679c25d438cc8412c2fa221cf1a824e6a
Status: Downloaded newer image for registry1-docker-io.repo.lab.pl.alcatel-lucent.com/library/alpine:latest
 ---> 14119a10abf4
Step 2/3 : RUN apk --no-cache add libcap && setcap cap_net_admin=eip /sbin/apk
 ---> Running in 923b46395907
fetch https://dl-cdn.alpinelinux.org/alpine/v3.14/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.14/community/x86_64/APKINDEX.tar.gz
(1/1) Installing libcap (2.50-r0)
Executing busybox-1.33.1-r3.trigger
OK: 6 MiB in 15 packages
Removing intermediate container 923b46395907
 ---> 97d6ec51d4b3
Step 3/3 : RUN setcap -v cap_net_admin=eip /sbin/apk
 ---> Running in 33260d42c77d
/sbin/apk: OK
Removing intermediate container 33260d42c77d
 ---> 5fc3c8660150
Successfully built 5fc3c8660150
Successfully tagged test:cap
```

**- Description for the changelog**
Fixed issue of file capabilities dropping when moving to next command in Dockerfile  during image building.
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

abdulrahimiliasu commented:
That's correct the Boolean should be set to ```true```  for those capabilities to be copied, but this change only makes sure that the 
"security-capability" attribute are copied. Everything else stay the same. But I do not know the specific reason for ```copyXattrs``` default to ```false```.

szubersk commented:
@thaJeztah I can't speak for @sargun but the feeling I get is that copying xattrs has been removed to speed copying up. d2b71b2 `Have VFS graphdriver use accelerated in-kernel copy` introduces two major optimizations: copy_file_range usage to transfer file content without switching from kernel- to userspace and back and xattrs ignoring.

`copy_file_range` can't take care of file attributes (or any other metadata) as it operates only on file content.
```
ssize_t copy_file_range(int fd_in, off64_t *off_in,
                        int fd_out, off64_t *off_out,
                        size_t len, unsigned int flags);
```

`copyXattr` requires one `getxattr(2)` call per file/directory + memory allocation (sometimes two) to store the result. No wonder @sargun was looking for an optimization there.
https://github.com/moby/moby/blob/c09789c1140a3b2e399b0c77c449b34d000e2362/daemon/graphdriver/copy/copy.go#L90-L91
https://github.com/moby/moby/blob/c09789c1140a3b2e399b0c77c449b34d000e2362/pkg/system/xattrs_linux.go#L8-L11

I'd lean towards suggestion of unconditionally enabling xattrs copying as current implementation is simply incorrect.

abdulrahimiliasu commented:
I've updated the commit  by changing copyXattrs to ```true``` instead of moving the block  https://github.com/abdulrahimiliasu/moby/blob/97ec8da56369452620dd1ad8ad51bfa2ccc622f1/daemon/graphdriver/copy/copy.go#L263-L265
out of
https://github.com/abdulrahimiliasu/moby/blob/97ec8da56369452620dd1ad8ad51bfa2ccc622f1/daemon/graphdriver/copy/copy.go#L258

szubersk commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -206,7 +206,7 @@ func DirCopy(srcDir, dstDir string, copyMode Mode, copyXattrs bool) error {
 		if err := os.Lchown(dstPath, int(stat.Uid), int(stat.Gid)); err != nil {
```
Changes in this file are irrelevant to the issue. Please consider removing.

sargun commented:
You should only copy the security.capability xattr in the VFS graphdriver, and not the trusted.overlay.opaque one.

szubersk commented:
@sargun @thaJeztah Is the PR in current form ok with you?

Commit:
Fix file capabilities droping in Dockerfile

doCopyXattrs() never reached due to copyXattrs boolean being false, as
a result file capabilities not being copied.

moved copyXattr() out of doCopyXattrs()

Signed-off-by: Illo Abdulrahim <abdulrahim.illo@nokia.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

szubersk commented:
@sargun @thaJeztah
Is there an issue that prevent merging this PR?

rightblank commented:
Hi, @thaJeztah, could you please help to look at if this could be merged? Thanks?

mikebrow reviewed (commented):
nit.. comments for the modified DirCopy() func and field name copyXattrs should be modified to reflect the boolean is just for switching off copying xattr trusted.overlay.opaque

mikebrow reviewed (approved):
LGTM functionally.. just the nit comment regarding variable naming and existing api // comments

thaJeztah commented:
I rebased the PR, renamed the `copyXattrs` argument to `copyOpaqueXattrs`, and updated the GoDoc for `CopyDir()`

thaJeztah reviewed (approved):
LGTM, thanks! And sorry for the long delay; this one dropped of my radar, and I saw the PR was still marked with "missing DCO", but that looks to have been resolved since.

mikebrow reviewed (approved):
LGTM
